### PR TITLE
`apply_operation` kernel for `MultiControlledX` on many wires

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
Currently, `default.qubit` applies the operation `MultiControlledX` by building a dense matrix.
This is rather expensive for large numbers of wires.

**Description of the Change:**
This PR adds a dispatch registration for `MultiControlledX` that falls back to `einsum` for less than eight wires in the operation.
For eight wires or more, it instead uses repeated `math.roll`, `math.transpose` and a single concatenation call to achieve the effect of the gate.

**Benefits:**
- Performance improvement on eight or more wires (see plots below).
- Makes more than 13 wires accessible for this operation.

Screenshot from 2023-10-13 16-56-16

Screenshot from 2023-10-13 16-54-24

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

### CODE REVIEW

**Review by timmysilv — COMMENTED** [review]
looks awesome! just one concern, then I'd say we're good to go

File: pennylane/devices/qubit/apply_operation.py [hunk 1]
```diff
@@ -263,6 +263,52 @@ def apply_cnot(op: qml.CNOT, state, is_state_batched: bool = False, debugger=Non
...
+            + [op.total_wires[-1]]
+            + op.total_wires[:-1].tolist()
+        )
+        + is_state_batched
+    )
+    # transpose_axes =  + is_state_batched for w in transpose_axes]
+    state = math.transpose(state, transpose_axes)
+
+    # Reshape the state into 3-dimensional array with axes [batch+other, target, controls]
+    state = state.reshape((-1, 2, 2 ** (len(op.wires) - 1)))
```
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:296]:
```suggestion
    state = math.reshape(state, (-1, 2, 2 ** (len(op.wires) - 1)))
```

Tensrflow tensors do not have a `reshape` method

File: pennylane/devices/qubit/apply_operation.py [hunk 2]
```diff
@@ -263,6 +263,52 @@ def apply_cnot(op: qml.CNOT, state, is_state_batched: bool = False, debugger=Non
...
+
+    # The part of the state to which we want to apply PauliX is now in the last entry along
+    # the third axis. Extract it, apply the PauliX along the target axis, and append a dummy axis
+    state_x = math.roll(state[:, :, -1], 1, 1)[:, :, np.newaxis]
+
+    # Stack the transformed part of the state with the unmodified rest of the state
+    state = math.concatenate([state[:, :, :-1], state_x], axis=2)
+
+    # Reshape into original shape and undo the transposition
+    state = math.transpose(state.reshape(orig_shape), np.argsort(transpose_axes))
```
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:306]:
```suggestion
    state = math.transpose(math.reshape(state, orig_shape), np.argsort(transpose_axes))
```

**Review by mudit2812 — COMMENTED** [review]
Looks great! I would suggest adding a small set of interface specific tests to catch errors like the ones I mentioned below :) Tensorflow will never not cause problems.

**Comment by dwierichs** [comment]
Thanks for the reviews @timmysilv  and @mudit2812 .
I fixed the Tensorflow support.
I also timed the other interfaces now, and of course the picture is much more complicated then...
image

I also am having a benchmarking discussion with @mlxd , and there are even more factors to take into account (like the number of threads).

Therefore my suggestion for this PR would be to only activate the dedicated kernel for a number of operation wires for which the dense matrix becomes too large. Of course this depends on the used hardware as well, but maybe it is sane to limit the matrix size to 2GB/14 wires ? For 15 wires it would be 8GB.
What do you think @mudit2812 , @mlxd, @timmysilv ?

**Comment by timmysilv** [comment]
honestly i'm finding it quite tough to read this graph 😅 what is the main point here? it seems like there is no discussion to be had beyond N=12 though

**Comment by dwierichs** [comment]
>honestly i'm finding it quite tough to read this graph 😅 what is the main point here?

Sorry, it is quite dense. The blue lines act with the `MultiControlledX` on all `N` wires. We see that this gains us a speedup (numbers below 1) in all interfaces, except for a single data point in the dotted line.
When the gate acts on a subset of more qubits (orange for first half, green for second half), though, it seems that there is some serious slow-down for numpy (dashed) and autograd (long dash-dot).

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -6,6 +6,9 @@
 
 <h3>Improvements 🛠</h3>
 
+* `default.qubit` no longer uses a dense matrix for `MultiControlledX` for more than 12 wires.
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:9]:
```suggestion
* `default.qubit` no longer uses a dense matrix for `MultiControlledX` for less than 12 device wires, or more than 13 operation wires
```

**Review by timmysilv — APPROVED** [review]
looks great! if you already have the benchmark code ready, could you show how this compares to master now? otherwise dw about it, this seems to be in accordance with what you found before

**Comment by mudit2812** [comment]
Happy to approve once the hiccups are resolved :)

**Comment by dwierichs** [comment]
Okay, so here is the updated timing:
image
I think I had accidentally included the fall-back in the previous timings, making them invalid.

I suggest the following branching now:
```python
if len(op.wires) < 9:
    # Fall back to einsum/tensordot. based on the standard criteria for that
return dedicated_kernel(op, state, is_state_batched)
```

File: pennylane/devices/qubit/apply_operation.py [hunk 4]
```diff
@@ -263,6 +264,69 @@ def apply_cnot(op: qml.CNOT, state, is_state_batched: bool = False, debugger=Non
...
+    if len_op < EINSUM_OP_WIRECOUNT_PERF_THRESHOLD and ndim < EINSUM_STATE_WIRECOUNT_PERF_THRESHOLD:
+        return apply_operation_einsum(op, state, is_state_batched)
+    if len_op < 9:
+        return apply_operation_tensordot(op, state, is_state_batched)
+    return _apply_multicontrolledx(op, state, is_state_batched)
+
+
+def _apply_multicontrolledx(op, state, is_state_batched):
+    r"""Apply MultiControlledX to state by composing transpositions, rolling of control axes
+    and the CNOT logic above."""
```
Comment (timmysilv) [inline pennylane/devices/qubit/apply_operation.py:286]:
```suggestion
    if len(op.wires) < 9:
        return _apply_operation_default(op, state, is_state_batched, debugger)
```

**Review by timmysilv — APPROVED** [review]
one suggestion to avoid duplicate logic, but otherwise lgtm!

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -9,6 +9,9 @@
 
 <h3>Improvements 🛠</h3>
 
+* `default.qubit` no longer uses a dense matrix for `MultiControlledX` for less than 12 device wires, or more than 13 operation wires
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:12]:
```suggestion
* `default.qubit` no longer uses a dense matrix for `MultiControlledX` for more than 8 operation wires.
```

File: pennylane/devices/qubit/apply_operation.py [hunk 6]
```diff
@@ -26,6 +26,7 @@
 
 EINSUM_OP_WIRECOUNT_PERF_THRESHOLD = 3
 EINSUM_STATE_WIRECOUNT_PERF_THRESHOLD = 13
+MATRIX_TOO_LARGE_THRESHOLD = 14
```
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:29]:
```suggestion
```

**Review by mudit2812 — COMMENTED** [review]
Two suggestions that seem to be artifacts from the old branching. Otherwise looks good. Happy to approve once addressed :)

**Review by mudit2812 — APPROVED** [review]
Thanks David!

**Pull request merged**